src/DeepCopy/CopyPolicy.cs

```diff
@@ -156,6 +156,40 @@ private Policy GetPolicy(Type type)
             return this.policies[type] = Policy.Mutable;
         }
 
+        public bool IsTrackingNecessary(Type type)
+        {
+            if (IsImmutable(type))
+            {
+                return false;
+            }
+            var duplicateCheck = new HashSet<Type> {type}; // if we encounter the root later on, tracking is necessary
+            return IsTrackingNecessary(type, duplicateCheck);
+        }
+
+        private bool IsTrackingNecessary(Type type, HashSet<Type> alreadyChecked)
+        {
+            if (type == typeof(object)) // we can't reason about object
+            {
+                return true;
+            }
+            var fields = GetCopyableFields(type);
+            foreach (var fieldInfo in fields)
+            {
+                var fieldType = fieldInfo.FieldType;                
+                var isImmutableField = IsImmutable(fieldType);
```

Comment:
I think immutability is the wrong thing to check here. An object graph might be immutable but still have cycles in it.

The check needs to be more along the lines of "can this field (or fields contained within) hold a reference to the root object."

If a field is of type object, the answer is yes (assuming you filter out ValueType at the top level), but more generally speaking the check should be `fieldType.IsAssignableFrom(rootType)`, where `rootType` is passed recursively to.

Reply:
Yes, I think you have a point there, I need to rethink that, my hashset was intended to check for cycles, but I think a different EqualityComparer needs to be used to get the IsAssignableFrom part working.

The problems I see with passing the root recursively is that it might need runtime checks and then the performance increase is gone anyway.
Need to make sure that there is no case where evaluating a Type returns NoTrackingNecessary, but if it is used in Parent object (list or similar) would return TrackingNecessary, as this would prevent us from caching NoTrackingNecessary.

Reply:
The tricky bit is that whether or not an object needs to be tracked depends on the root object which contains it (which of course can vary).

So the question becomes 'can multiple references to a given object of type X appear in an object graph whose root is type Y'. That means that you may have to generate multiple copy routine for each type X.

Reply:
Yes exactly, the "simple" approach is to not reuse any copy routines anymore but generate a unique one for each unique object graph.